File: src/com/jetbrains/idear/recognizer/AudioInputStreamWithAdjustableGain.java
package com.jetbrains.idear.recognizer;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.TargetDataLine;
import java.io.IOException;
import java.io.InputStream;

public class AudioInputStreamWithAdjustableGain extends AudioInputStream {

    private static final double DEFAULT_MASTER_GAIN = 1.0;

    private double masterGain;

    public AudioInputStreamWithAdjustableGain(InputStream stream, AudioFormat format, long length) {
        super(stream, format, length);

        masterGain = DEFAULT_MASTER_GAIN;
    }

    public AudioInputStreamWithAdjustableGain(TargetDataLine line) {
        super(line);

        masterGain = DEFAULT_MASTER_GAIN;
    }

    @Override
    public int read() throws IOException {
        return super.read();
    }

    @Override
    public int read(byte[] b) throws IOException {
        int read = super.read(b);
        for (int i = 0; i < read; ++i) {
            b[i] = adjust(b[i]);
        }
        return read;
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        int read = super.read(b, off, len);
        for (int i = off; i < off + read; ++i) {
            b[i] = adjust(b[i]);
        }
        return read;
    }

    public double setMasterGain(double mg) {
        double pmg = masterGain;
        masterGain = mg;
        return pmg;
    }

    private byte adjust(byte b) {
        return (byte) Math.min(b * masterGain, Byte.MAX_VALUE);
    }
}

File: src/main/java/org/openasr/idiolect/recognizer/AudioInputStreamWithAdjustableGain.kt
package org.openasr.idiolect.recognizer

import java.io.IOException
import java.lang.Byte.*
import javax.sound.sampled.*

class AudioInputStreamWithAdjustableGain internal constructor(line: TargetDataLine) : AudioInputStream(line) {
    private val DEFAULT_MASTER_GAIN = 1.0
    private val DEFAULT_NOISE_LEVEL = 0.0
    private val MAX_SAMPLE_VALUE = 32768.0
    private var masterGain = DEFAULT_MASTER_GAIN
    private var noiseLevel = DEFAULT_NOISE_LEVEL

//    private val filter = NoiseSuppressor()
//
//    private val frameSize = 256
//    private val frameBuffer = IntArray(frameSize)


    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        val bytesRead = super.read(b, off, len)
        if (bytesRead == -1) {
            return bytesRead
        }

        val end = off + bytesRead

        // step by 2 bytes because `sampleSizeInBytes = format.sampleSizeInBits / 8` = 2
        for (i in off until end step 2) {
            // TODO: delegate to Filter.filterSample(b, i) and move into recognizer.filter
            var sample = ((b[i + 1].toInt() shl 8) + (b[i].toInt() and 0xff)) // / MAX_SAMPLE_VALUE

            val amplifiedSample = filterSample(sample)

            b[i] = (amplifiedSample and 0xff).toByte()
            b[i + 1] = ((amplifiedSample shr 8) and 0xff).toByte()
        }

        return bytesRead
    }

   /* @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        val bytesRead = super.read(b, off, len)
        if (bytesRead == -1) {
            return bytesRead
        }

        val end = off + bytesRead
        var j = 0

        // step by 2 bytes because `sampleSizeInBytes = format.sampleSizeInBits / 8` = 2
        for (i in off until end step 2) {
            // TODO: delegate to Filter.filterSample(b, i) and move into recognizer.filter
            var sample = ((b[i + 1].toInt() shl 8) + (b[i].toInt() and 0xff)) // / MAX_SAMPLE_VALUE

            frameBuffer[j++] = sample

            if (j == frameSize) {
                // frameBuffer is full, filter and update `b` buffer
                filter.filterSamples(frameBuffer, j)

                for (k in 0 until j) {
                    val filteredSample = frameBuffer[k]
                    val l = off + (k shl 1)

                    b[l] = (filteredSample and 0xff).toByte()
                    b[l + 1] = ((filteredSample shr 8) and 0xff).toByte()
                }

                j = 0
            }
        }

        return bytesRead
    }*/

    fun filterSample(sample: Int): Int {
        if (sample < noiseLevel && sample > -noiseLevel) {
            return 0
        }
        return (sample * masterGain).toInt().coerceIn(-32768, 32767)
    }

    fun setMasterGain(mg: Double) = mg.let { masterGain = it }

    fun setNoiseLevel(nl: Double) = nl.let { noiseLevel = it }

    private fun dump(b: ByteArray, off: Int, len: Int) {
        val sb = StringBuilder()
        for (i in off until off + len - 1) {
            sb.append(b[i].toInt())
            if (i != off + len - 1)
                sb.append(", ")
        }
    }

    private fun adjust(b: Byte) {
        cut((b * masterGain).toInt().toByte())
    }

    private fun cut(b: Byte) =
        if (b < MAX_VALUE * noiseLevel && b > MIN_VALUE * noiseLevel) 0 else b
}
